code/controllers/subsystem/discord.dm

```diff
@@ -81,70 +91,186 @@ SUBSYSTEM_DEF(discord)
+		one_time_token = trim("[pick(common_words)]-[pick(common_words)]-[pick(common_words)]-[pick(common_words)]-[pick(common_words)]-[pick(common_words)]", 100)
+
+		not_unique = find_discord_link_by_token(one_time_token, timebound = TRUE)
+
+	// Insert into the table, null in the discord id, id and timestamp and valid fields so the db fills them out where needed
+	var/datum/db_query/query_insert_link_record = SSdbcore.NewQuery(
+		"INSERT INTO [format_table_name("discord_links")] (ckey, one_time_token) VALUES(:ckey, :token)",
+		list("ckey" = ckey_for, "token" = one_time_token)
 	)
-	if(!query_get_discord_ckey.Execute())
-		qdel(query_get_discord_ckey)
-		return
-	if(query_get_discord_ckey.NextRow())
-		. = query_get_discord_ckey.item[1]
-	qdel(query_get_discord_ckey)
-
-// Finalises link
-/datum/controller/subsystem/discord/proc/link_account(ckey)
-	var/datum/db_query/link_account = SSdbcore.NewQuery(
-		"UPDATE [format_table_name("player")] SET discord_id = :discord_id WHERE ckey = :ckey",
-		list("discord_id" = account_link_cache[ckey], "ckey" = ckey)
+
+	if(!query_insert_link_record.Execute())
+		qdel(query_insert_link_record)
+		return ""
+
+	//Cleanup
+	qdel(query_insert_link_record)
+	return one_time_token
+
+/**
+  * Find discord link entry by the passed in user token
+  *
+  * This will look into the discord link table and return the *first* entry that matches the given one time token
+  *
+  * Remember, multiple entries can exist, as they are only guaranteed to be unique for their validity period
+  *
+  * Arguments:
+  * * one_time_token the string of words representing the one time token
+  * * timebound A boolean flag, that specifies if it should only look for entries within the last 4 hours, off by default
+  *
+  * Returns a [/datum/discord_link_record]
+  */
+/datum/controller/subsystem/discord/proc/find_discord_link_by_token(one_time_token, timebound = FALSE)
+	var/timeboundsql = ""
+	if(timebound)
+		timeboundsql = "AND timestamp >= Now() - INTERVAL 4 HOUR"
+	var/query = "SELECT CAST(discord_id AS CHAR(25)), ckey, MAX(timestamp), one_time_token FROM [format_table_name("discord_links")] WHERE one_time_token = :one_time_token [timeboundsql] GROUP BY ckey, discord_id, one_time_token LIMIT 1"
+	var/datum/db_query/query_get_discord_link_record = SSdbcore.NewQuery(
+		query,
+		list("one_time_token" = one_time_token)
 	)
-	link_account.Execute()
-	qdel(link_account)
-	account_link_cache -= ckey
-
-// Unlink account (Admin verb used)
-/datum/controller/subsystem/discord/proc/unlink_account(ckey)
-	var/datum/db_query/unlink_account = SSdbcore.NewQuery(
-		"UPDATE [format_table_name("player")] SET discord_id = NULL WHERE ckey = :ckey",
+	if(!query_get_discord_link_record.Execute())
+		qdel(query_get_discord_link_record)
+		return
+	if(query_get_discord_link_record.NextRow())
+		var/result = query_get_discord_link_record.item
+		. = new /datum/discord_link_record(result[2], result[1], result[4], result[3])
```

Comment:
Since this is only used in `generate_one_time_token()` to test the `while(not_unique)` loop, couldn't you just set `. = TRUE` and skip having to create a datum that isn't actually used for anything?

Reply:
it will be used in later code that lets people manage discord links in game.